src/test/java/lambda/part2/exercise/Exercise2.java

```diff
@@ -0,0 +1,108 @@
+            Predicate<Person> right) {
+        return (p) -> left.test(p) && right.test(p);
+    }
+
+    @Test
+    public void personHasNotEmptyLastNameAndFirstNameUsingLogicalOperators() {
+        Predicate<Person> personHasEmptyFirstName = (p) -> p.getFirstName().isEmpty();
+        Predicate<Person> personHasEmptyLastName = (p) -> p.getLastName().isEmpty();
+
+        Predicate<Person> personHasNotEmptyFirstName = negateUsingLogicalOperator(
+                personHasEmptyFirstName);
+        Predicate<Person> personHasNotEmptyLastName = negateUsingLogicalOperator(
+                personHasEmptyLastName);
+
+        Predicate<Person> personHasNotEmptyLastNameAndFirstName = andUsingLogicalOperator(
+                personHasNotEmptyFirstName, personHasNotEmptyLastName);
+
+        assertTrue(
+                personHasNotEmptyLastNameAndFirstName.test(new Person("Алексей", "Доренко", 40)));
+        assertFalse(personHasNotEmptyLastNameAndFirstName.test(new Person("Николай", "", 30)));
+        assertFalse(personHasNotEmptyLastNameAndFirstName.test(new Person("", "Мельников", 20)));
+    }
+
+    // TODO метод (T -> boolean) -> (T -> boolean)
+    // TODO - возвращает новый предикат, являющийся отрицанием исходного
+    // TODO - при реализации использовать логический оператор !
+    private <T> Predicate<T> negate(Predicate<T> predicate) {
+        return (t) -> !predicate.test(t);
+    }
+
+    // TODO метод (T -> boolean, T -> boolean) -> (T -> boolean)
+    // TODO - возвращает новый предикат, объединяющий исходные с помощью операции "AND"
+    // TODO - при реализации использовать логический оператор &&
+    private <T> Predicate<T> and(Predicate<T> left, Predicate<T> right) {
+        return (t) -> left.test(t) && right.test(t);
+    }
+
+    @Test
+    public void personHasNotEmptyLastNameAndFirstNameUsingGenericPredicates() {
+        Predicate<Person> personHasEmptyFirstName = (p) -> p.getFirstName().isEmpty();
+        Predicate<Person> personHasEmptyLastName = (p) -> p.getLastName().isEmpty();
+
+        Predicate<Person> personHasNotEmptyFirstName = negate(personHasEmptyFirstName);
+        Predicate<Person> personHasNotEmptyLastName = negate(personHasEmptyLastName);
+
+        Predicate<Person> personHasNotEmptyLastNameAndFirstName = and(personHasNotEmptyFirstName,
+                personHasNotEmptyLastName);
+
+        assertTrue(
+                personHasNotEmptyLastNameAndFirstName.test(new Person("Алексей", "Доренко", 40)));
+        assertFalse(personHasNotEmptyLastNameAndFirstName.test(new Person("Николай", "", 30)));
+        assertFalse(personHasNotEmptyLastNameAndFirstName.test(new Person("", "Мельников", 20)));
+    }
+
+    @Test
+    public void personHasNotEmptyLastNameAndFirstNameUsingStandardMethods() {
+        Predicate<Person> personHasEmptyFirstName = (p) -> p.getFirstName().isEmpty();
+        Predicate<Person> personHasEmptyLastName = (p) -> p.getLastName().isEmpty();
```

Comment:
I would suggest extracting `Predicate<Person> personHasEmptyFirstName` and `Predicate<Person> personHasEmptyLastName`, so you don't have to declare them in every test.